packages/nitro-protocol/contracts/examples/EmbeddedApplication.sol

```diff
@@ -0,0 +1,438 @@
+// 2a. A and I remove $G_{AI}$ and absorb the outcome of J into their ledger channel
+// 2b. B and I remove $G_{BI}$ and absorb the outcome of J into their ledger channel
+
+// On chain challenge path:
+//
+// 1. J outcome recorded on chain. Each A and B can unilaterally progress the channel once.
+// (So the the latency of finalizing a channel on chain is at most 6 timeout periods. In practice, since I never needs to update the channel, the latency is likely 3 timeout periods.)
+// 2. Claim with $G_{AI}$ is called.
+// 3. Claim with $G_{BI}$ is called. After this, all participants have pulled out their funds.
+
+/**
+ * @dev The EmbeddedApplication contract embeds a subchannel X.
+ */
+contract EmbeddedApplication is
+    IForceMoveApp2 // We need a new interface to allow signedBy information into the validTransition function
+{
+    // 2-party SupportProof
+    struct SupportProof {
+        IForceMove.FixedPart fixedPart;
+        VariablePart[] variableParts; // either one or two states
+        uint48 turnNumTo;
+        IForceMove.Signature[2] sigs; // one for each participant
+        uint8[2] whoSignedWhat; // whoSignedWhat = [0,0] or [0,1] or [1,0]
+    }
+
+    // Finite states for J
+    enum AlreadyMoved {A, B, AB, ABI}
+
+    // Application Data for J
+    struct AppData {
+        bytes32 channelIdForX;
+        SupportProof supportProofForX;
+        AlreadyMoved alreadyMoved;
+    }
+
+    function validTransition(
+        VariablePart memory from,
+        VariablePart memory to,
+        uint48, // turnNumTo (unused)
+        uint256, // nParticipants (unused)
+        uint256, // signedByFrom (unused) - Who has signed the "from" state?
+        uint256 signedByTo // Who has signed the "to" state?
+    ) public override pure returns (bool) {
+        // parameter wrangling
+        bool signedByA = ForceMoveAppUtilities.isSignedBy(signedByTo, 0);
+        bool signedByB = ForceMoveAppUtilities.isSignedBy(signedByTo, 1);
+        AppData memory fromAppData = abi.decode(from.appData, (AppData));
+        AppData memory toAppData = abi.decode(to.appData, (AppData));
+        Outcome.AllocationItem[] memory fromAllocation = decode3PartyAllocation(from.outcome);
+        Outcome.AllocationItem[] memory toAllocation = decode3PartyAllocation(to.outcome);
+
+        require(
+            fromAllocation[0].destination == toAllocation[0].destination &&
+                fromAllocation[1].destination == toAllocation[1].destination &&
+                fromAllocation[2].destination == toAllocation[2].destination,
+            'destinations may not change'
+        );
+
+        require(fromAllocation[2].amount == toAllocation[2].amount, 'p2.amt !constant');
+        require(
+            fromAllocation[0].amount + fromAllocation[1].amount + fromAllocation[2].amount ==
+                toAllocation[0].amount + toAllocation[1].amount + toAllocation[2].amount,
+            'total allocation changed'
+        );
```

Comment:
I prefer to use constants, rather than "magic numbers", for indexes of importance.

As an example benefit, if we wish to change the outcome ordering, we don't need to find-and-replace all uses of the number "1".

Reply:
I think we could get away with an enum like
```
enum indices {
  A,
  B,
  I
}
```

While this might technically work, it is a bit "clever", since it depends on the enum's values being sequential integers starting at 0.

Reply:
1a69e8c